application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/DefaultLicensedExtensionManager.java

```diff
@@ -60,6 +60,11 @@
      */
     public static final String LICENSOR_EXTENSION_ID = "com.xwiki.licensing:application-licensing-licensor-api";
 
+    /**
+     * Cache the list of installed licensed extensions that are not covered by the license of another extension.
+     */
+    private Set<ExtensionId> cachedMandatoryLicensedExtensions;
```

Comment:
@mflorea Would be the right fix to use `@InstantiationStrategy(ComponentInstantiationStrategy.PER_LOOKUP)`? This would make possible to add a state.
Or it is better to maybe use a cache component for saving the extensions list?

Reply:
If ``DefaultLicensedExtensionManager`` stops being a singleton then the cache doesn't matter that much anymore. Each component that injects ``LicensedExtensionManager`` will have its own instance with its own cache, which basically means that the list of mandatory licensed extensions is going to be recomputed each time.

I think it needs to remain a singleton but you need to make sure that whenever its state (``cachedMandatoryLicensedExtensions``) is read (``getMandatoryLicensedExtensions``) the return value is valid. The state can become invalid through writes, so you need to make sure that after each write the state is valid. If multiple writes are needed to go from one valid state to another valid state then you need to make sure those writes cannot be interleaved with a read from a different thread. This is usually done using synchronized blocks/methods or semaphores.

In you case the problems are:
* ``cachedMandatoryLicensedExtensions`` can be in a partial state when read (because some other thread is computing the cache) -> you can fix this by setting ``cachedMandatoryLicensedExtensions`` only at the end when you have the final list of mandatory licensed extensions
* multiple threads can compute the cache at the same time; only one thread should be allowed to do this; you can move the code that computes the cache in a **synchronized** method
* some thread can invalidate the cache (set to null) while another one is computing the case; you can fix this with the first item above + returning the computed value instead of ``this.cachedMandatoryLicensedExtensions`` in ``getMandatoryLicensedExtensions``. Basically:
  ```
  Set<ExtensionId> cache = computeMandatoryLicensedExtensions();
  this.cachedMandatoryLicensedExtensions = cache;
  return cache;
  ```